.github/actions/python-maturin/pre-merge/action.yml

```
@@ -0,0 +1,178 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: python-maturin-pre-merge
+description: Python pre-merge testing with maturin github iggy actions
+
+inputs:
+  task:
+    description: "Task to run (lint, test, build)"
+    required: true
+
+runs:
+  using: "composite"
+  steps:
+    - name: Setup Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.11"
+
+    - name: Setup Rust with cache
+      uses: ./.github/actions/utils/setup-rust-with-cache
+      with:
+        cache-targets: false  # Only cache registry and git deps, not target dir (sccache handles that)
+
+    - name: Cache pip
+      uses: actions/cache@v4
+      with:
+        path: ~/.cache/pip
+        key: pip-${{ runner.os }}-${{ hashFiles('foreign/python/pyproject.toml') }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip wheel setuptools
+
+        # Install maturin for building
+        pip install 'maturin>=1.2,<2.0'
+
+        # Install testing dependencies from pyproject.toml
+        cd foreign/python
+        pip install -e ".[testing,examples,dev]"
+        cd ../..
+
+        # Install additional CI dependencies for Docker testing if needed
+        if [ "${{ inputs.task }}" = "test" ]; then
+          pip install 'testcontainers[docker]>=3.7.0,<5.0' || true
+        fi
+      shell: bash
+
+    - name: Lint and format check
+      if: inputs.task == 'lint'
+      run: |
+        # Build list of directories to check using absolute paths
+        DIRS_TO_CHECK="${GITHUB_WORKSPACE}/foreign/python/tests"
+        DIRS_TO_CHECK="$DIRS_TO_CHECK ${GITHUB_WORKSPACE}/foreign/python/python_examples"
+        DIRS_TO_CHECK="$DIRS_TO_CHECK ${GITHUB_WORKSPACE}/examples/python/getting-started"
+
+        STUB_FILE="${GITHUB_WORKSPACE}/foreign/python/apache_iggy.pyi"
+
+        echo "Directories to check: $DIRS_TO_CHECK"
+        echo "Stub file: $STUB_FILE"
+
+        echo "ruff check --select I $DIRS_TO_CHECK"
+        ruff check --select I $DIRS_TO_CHECK
+        echo "ruff version: $(ruff --version)"
+
```

```suggestion

        echo "ruff format $DIRS_TO_CHECK"
        ruff format "$DIRS_TO_CHECK"
```

This runs the ruff formatter, which is independent of the ruff linter (ruff check).

---

it doesn't really work. when quoted, ruff treats it as a single path with spaces in it, which doesn't exist.